lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -897,59 +901,106 @@ _evaluateMinimumFee
     :: Cardano.ProtocolParameters
     -> SealedTx
     -> Maybe Coin
-_evaluateMinimumFee pp tx =
-    fromCardanoLovelace <$> minFee
+_evaluateMinimumFee pp tx = do
+    anyShelleyTx <- inAnyShelleyBasedEra (cardanoTx tx)
+    pure $ withShelleyBasedBody anyShelleyTx $ \_era body ->
+        fromCardanoLovelace $
+            Cardano.evaluateTransactionFee
+                pp
+                body
+                (estimateNumberOfWitnesses body)
+                0
+
+-- | Estimate the size of the transaction (body) when fully signed.
+_estimateSignedTransactionSize
+    :: Cardano.IsShelleyBasedEra era
+    => Cardano.ProtocolParameters
+    -> Cardano.TxBody era
+    -> TxSize
+_estimateSignedTransactionSize pparams body =
+    let
+        nWits = estimateNumberOfWitnesses body
+
+        -- Hack which allows us to rely on the ledger to calculate the size of
+        -- witnesses:
+        feeOfWits = minfee nWits - minfee 0
+        sizeOfWits =
```

Comment:
I reckon it might be nice to add some type signatures to `feeOfWits` and `sizeOfWits`.

What do you think?

Reply:
Sure 👍 -> 4c9e97735a918ba06dedf4f5059df5c22c4e0853

Reply:
Although the usage of `Natural` representing Coin — which makes the `feeOfWits `quotRem` feePerByte` readable — makes the type-signatures not that helpful 🤷‍♂️

Reply:
> Although the usage of `Natural` representing Coin — which makes the `feeOfWits 'quotRem' feePerByte` more readable — makes the type-signatures less helpful man_shrugging

I've added a suggestion here:
https://github.com/input-output-hk/cardano-wallet/pull/3166#discussion_r826650333